### ISSUE TITLE
Standartisation of domain specific properties prefixes and schemas

### ISSUE THREAD

**Issue by osdm** [editorconfig/editorconfig#332]
Currently I know about two usages of domain specific properties: .NET compiler+Visual Studio and ReSharper+Rider. I think there are more usages. There are two problems with them:
1. Currently two editors can use domain specific properties with the same name, but different semantics.
2. There is no way to get a lists of domain specific properties from different tools in order to validate and provide code completion for .editorconfig files.

My suggestions:
1. Let's create a standard (JSON?) format for describing domain specific properties
2. All domain specific properties should have a prefix (like csharp_, dotnet_ or resharper_), and those prefixes should be registered in some common github file (again JSON?) along with a link to a file with properties description (also could be maintained in domain-specific github repos)

**Comment by xuhdev** [comment]
I don't quite understand what the JSON part means. Can you illustrate?

**Comment by osdm** [comment]
JSON is a file format. I suggest using it for describing prefixes for domain specific properties, for example:

    [
      {
        "prefix": "csharp_",
        "description": "Code style properties for C# language",
        "schema_url": "https://raw.githubusercontent.com/dotnet/someneweditorconfigrepository/master/propertylist.json"
      },
      {
        "prefix": "resharper_",
        "description": "Code style and inspection properties supported by ReSharper and Rider", 
        "schema_url": "https://raw.githubusercontent.com/jetbrains/someneweditorconfigrepository/master/resharperpropertylist.json"
      }
    ]

Schema files (property list) can look like this, for example:

    [
      {
        "name": "resharper_place_something_on_new_line",
        "description": "Denotes whether a line break should be inserted before Something",
        "possible_values": "boolean" 
      },
      {
        "name": "resharper_blank_lines_before_something",
        "description": "Ensures that at least a specified number of blank lines is present before Something",
        "possible_values": "int" 
      },
      {
        "name": "resharper_brace_style",
        "description": "Denotes a brace style",
        "possible_values": [
          { "name": "end_of_line", "description": "Places opening brace at the end of the line, one indent inside braces, no indent for closing brace" },
          { "name": "next_line", "description": "Places opening brace on a new line, no indent for braces, one indent inside braces" }
        ] 
      }
    ]

**Comment by osdm** [comment]
P.S. If you don't like JSON, INI format or YAML could be used instead.

**Comment by xuhdev** [comment]
I'm a bit confused. I thought you are suggesting some format compliance in the editorconfig files for different editors when they specify their editor specific options. Are you suggesting that we put this information in JSON format, and automatically generate a webpage to describe them? But I think this should be a responsibility for the editor developers to explain in their doc.

**Comment by osdm** [comment]
I'm suggesting both. Explaining in a doc is not sufficient, because it doesn't allow to develop a tool to provide a validation and code assistance when editing .editorconfig files. For example, see a screenshot from the linked project https://github.com/madskristensen/EditorConfigLanguage/blob/master/art/keyword-intellisense.png . You may notice that resharper properties are absent here, only csharp and dotnet domain specific properties are there. Currently there is no way for such project to get all possible properties from all domains, except for manually scavenging all docs and tracking all updates in them (and that is a lot of work and nobody would do it).

And one more thing. You say "editor specific options". But they are not "editor specific". They are "domain specific". For example, "csharp_*" properties are supported by both Visual Studio and Rider IDEs, so they are not editor specific, but language specific. Well, given that C# is a Microsoft language, Microsoft is responsible for providing docs. But what about "javascript_*" properties? Or "python_*" properties? Currently they are out of scope for editorconfig project. Separate open source projects could be established for discussing and standardization of such properties and providing plugins and conversion tools. This proposal provides a way to track such projects and make them "semi-official" in editorconfig world.

**Comment by xuhdev** [comment]
@osdm Ah I see. For some reasons I misunderstood it as editor-specific options. OK, now I agree with you: We should do this.

The difficulty now is that, if we provide such a list, how many editors will implement them? This can be misleading for developers who read that page. Perhaps, we can at least initially say that these are "experimental" and "are not widely supported yet by many editors/only supported by a few editors".

For this, I believe a voting is required, since it somewhat affects the file format. @editorconfig/board-member  What do you think?

**Comment by ffes** [comment]
In general I like the idea. The file format in json is fine with me. It is what we have chosen for the defaults repo as well.

We should make sure that all those json formats are similar to each other and that they have validation schemas.

Are third party generators of additional domains willing to create a schema?

Before we start a vote we need to figure those things out s bit more.

**Comment by osdm** [comment]
I'll definitely create schemas for ReSharper. I can also create schemas for C# and .NET if Microsoft won't do it.

**Comment by sindresorhus** [comment]
EditorConfig is for "editor config", not code style. Code style should be handled by language-specific linters.

**Comment by osdm** [comment]
@sindresorhus Sorry, but many editors (IDEs) have ability to check and enforce code style (for example via builtin formatter). Devs need a way to sync code style between different IDE installations and even between different IDEs, so why not EditorConfig? Microsoft's domain specific properties for C# are already used to sync code style between Visual Studio and Rider IDEs.

Besides, EditorConfig already defines essential parts of code style (indent size and style). It is very inconvenient to have parts of code style defined in EditorConfig and parts defined by other means.

**Comment by sindresorhus** [comment]
> so why not EditorConfig?

Why not define it in `.gitignore` then? Same argument. Because it overloads `.editorconfig` for something it was not intended for. Starting to support general code style is a slippery slope.

**Comment by osdm** [comment]
@sindresorhus 
My previous answer already contains two arguments about why code style is a part of, well, "editor config" and so a part of EditorConfig as intended and not .gitignore.

Besides, let me quote a first sentence from editorconfig.org:
> EditorConfig helps developers define and maintain consistent **coding styles** between different editors and IDEs.

Can you tell me where slippery slope that you mentioned leads to?

**Comment by sindresorhus** [comment]
All of the current official EditorConfig rules are already built-in settings in most editors. The things you're proposing require language parsing.

**Comment by osdm** [comment]
@sindresorhus Maybe there is some misunderstanding. I don't propose to add anything that require language parsing to "official EditorConfig rules". All I'm proposing is a way to maintain a list of domain specific properties that people are adding for use in their specific domains (for example, for sharing settings between editors that can do language parsing). Please note, that [EditorConfig wiki](https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#ideas-for-domain-specific-properties) already suggest properties that require language parsing (curly_bracket_next_line, spaces_around_operators, indent_brace_style), so that isn't something new.

**Comment by ppalaga** [comment]
I very much like this proposal. I came across the same idea when thinking about how to effectively sync formatting rules across IDEs and build tools for Java. `.editorconfig` files are a good fit for the task. I see the same problem as @osdm : having every language specific rule to pass the voting here at the EditorConfig level would be very impractical. What we need is an offcial EditorConfig support for property namespaces. In other words, we need to introduce a way how can it be decided for a given property if it is either 

(a) general - its semantics is defined by EditorConfig or

(b) domain specific - its semantics is defined by the owner of the domain namespace

As for technical solution, @osdm proposes simple prefixes like `csharp_`, `dotnet_` or `resharper_` and a remote repository where these prefixes would be registered. I like both the idea of the prefix and the repo with a minor modification: The prefix should be delimited by some special character, so that domain specific properties can be identified offline by purely syntactic means.

Clearly, we cannot use colon, equals and hash sign, because they already have a meaning. Underscore is not practical because it is widely used as a general purpose word separator. By quickly looking at the parser, say, slash `/` could be a good namespace prefix separator. E.g. 

```
java/my_java_specific_property = 42
```

Namespace unaware tools would swallow such properties as syntactically correct but they would ignore them as unknown, while namespace aware tools could do the namespace mapping a bit more effectively. WDYT?

**Comment by holandaleo** [comment]
What I can see in the future is, if debate starts to be so strict about limitations of editorconfig, what will happen is that someone will create another project with another name (let's say, EditStyleConfig) that standardize the way IDEs and code editors formats language specific things. But then, that will include the basics such as space/tabs, indentation etc and will end up killing EditorConfig project. 

> EditorConfig is for "editor config", not code style. Code style should be handled by language-specific linters.

By the way, this is the first statement extracted from https://editorconfig.org/ in _What is EditorConfig? section_:

> EditorConfig helps developers define and maintain consistent *coding styles* between different editors and IDEs.

**Comment by bukajsytlos** [comment]
Is this project dead?

**Comment by MisinformedDNA** [comment]
@bukajsytlos Maybe not "dead", but not really evolving either.

**Comment by florianb** [comment]
@bukajsytlos - thanks for your interest. This project is not dead and in fact we're currently working on fundamental changes to the organization of the EditorConfig standard itself to become more flexible handling extensions.

**Comment by bukajsytlos** [comment]
Thanks, good to hear. Because from outside it looks like there is not much activity, lot of issues, no PRs and this issue is so fundamental and 2 years old.
Good luck moving forward with this useful project.

**Comment by florianb** [comment]
Thanks - i am sorry for that impression. This is the main repository for issues people having with EditorConfig, the most of the work aka PRs is happening in other repositories. Due to the amount of users, involved architectures and environments things move slowly to avoid breaking changes. But we're on it. And things should hopefully become partially more agile soon.

**Comment by osdm** [comment]
@florianb  I think you are more likely to break things by moving slowly, because domain-specific properties are already multiplying in different projects. Visual Studio has "dotnet_" and "csharp_" properties, Intellij IDEA has "ij_", ReSharper/Rider has "resharper_", and people are already using them. You'll have to accept them into standard "as is" to avoid breakage.

**Comment by florianb** [comment]
@osdm - i disagree, while it is currently not explicitly defined how the EditorConfig standard treats custom properties, they are absolutely legit and i'd embracing using them.

I think we shouldn't standardize everything we probably could standardize. We take care of a solid and backwards compatible foundation every user might build on. I currently don't see a strong requirement justifying the investment of resources in creating and maintaining third-party namespaces. This is something third-party developers can perfectly take care for.

There is work underway formalizing the aspect of how the EditorConfig may be used with custom properties in a safe way. And i can imagine we might make the use of custom properties more visible.

**Comment by cxw42** [comment]
Late to the party, but ---

I propose a vote on:
- Adding property namespaces with / to the specification per https://github.com/editorconfig/editorconfig/issues/332#issuecomment-391255016
- Adding a file to the specification repo to hold the list of registered domain-specific properties
- Adding the schema of the file to the specification
- Opening registration on a first-come, first-served basis via PRs to the file
- Documenting the formal requirements of such a PR in the specification
- (new) Using YAML for the file (because JSON doesn't allow comments in its normal incarnation, and I miss them every time I have to work with a JSON config file)

Would such a vote meet the needs expressed in this issue?

**Comment by florianb** [comment]
I think i can remember we already discussed the "list of registered domain-specific properties" in another thread - i would like to avoid that to avoid putting us in responsibility.

The first-come-first-serve approach wouldn't work anyways in my opinion, since it would require every possible stakeholder to actually recognizing our opening on that case.

I'd prefer a more liberal approach, projects should just be able to announce private-properties (which are just properties containing a slash).

Sure that would add potential for conflicts, but it's hard enough to pick up on highly demanded core-transformations, we shouldn't overwhelm the board with issues originated at third-party projects as well.

**Comment by florianb** [comment]
Oh and as an additional idea we could have a formal "spec"-file where developers are able to describe their custom properties in a formal way, so that one could aggregate those information, let's say in a search or a help-system.

**Comment by cxw42** [comment]
@florianb We discussed a registered list of languages at (among others) #190 and https://github.com/editorconfig/editorconfig/issues/404#issuecomment-533028163 .  This also relates to #415.

- A first-come-first-served registry would not require Board action --- anyone with rights to merge a PR could handle a request.  
  - We could put the file in a separate repo with broader push access if that would help.
- I don't think advertising is a problem --- people are beating down our door in this repo asking for new properties :) .  If we pinned an issue that said "Custom-property requests go here", I think we would cut down on issue traffic in this repo quite a bit.
- I agree about the spec file --- that was my understanding of the mention above of a schema.  https://github.com/editorconfig/editorconfig/issues/332#issuecomment-355812750

I just realized some of the discussion above was about _prefixes_ --- I'd suggest a registry of prefixed _properties_ instead.  That way developers of editors can have one place to look for all property information.

**Comment by osdm** [comment]
IMHO, it's too late to have slash '/' to denote prefixes. Although it's a very good idea, I don't think everybody who is already using domain-specific properties would rush to change their prefixes. We are stuck with using '_' for backward compatibility.

**Comment by cxw42** [comment]
@osdm from the standpoint of the _specification_, there are no backwards-compatibility guarantees for non-standard properties.  I agree that editors that already use non-standard properties will probably continue to do so.  However, I don't think that prevents the specification from moving forward.

**Comment by florianb** [comment]
@cxw42 - ah thank you very much those threads i was searching for.

> A first-come-first-served registry would not require Board action --- anyone with rights to merge a PR could handle a request.

How would conflicts being solved then? Probably it would be a good idea to (at least) recommend using domain names as property-prefix?

>  pinned [...] issue that said "Custom-property requests go here"

Sure, that would work, but we should expect major players to show up after somebody else already "claimed" a prefix (especially if it's entirely free how to reserve them).

> I agree about the spec file

Thanks for the reference - missed that one.

> I'd suggest a registry of prefixed properties instead.

I agree - as already said, i'd really favor a system letting the board "just" supervising conflicts and edge cases, without involving too much basic interaction.

---

> We are stuck with using '_' for backward compatibility.

@osdm - in my opinion the underline-delimited properties remain as valid as they are now, unless the board decides once to expand a core-rule to overlap with one of those self chosen property-names. I guess that's unlikely to happen.
At the end i hope people just adopt the new naming scheme for their custom-properties.

**Comment by cxw42** [comment]
@florianb re. conflicts, if it's purely first-come-first-served, there will be no conflicts.  The earliest commit timestamp wins.  If we require PRs or proposals include enough information to make a decision, we should be able to close PRs that are squatting, violate terms of service, or are otherwise undesirable.

As far as prefixes, I agree with using domain names (preferably reversed) as prefixes.  Language names could also be used as prefixes.

**Comment by florianb** [comment]
Thanks - i'm ready to go ahead if nobody complaints.

**Comment by ppalaga** [comment]
> I'd suggest a registry of prefixed properties instead.

I proposed having just a registry of prefixes so that communities around the individual prefixes can coin their own rules in their own GH orgs (or similar) how the set of properties in their namespaces should evolve. 

If the registry under the editoconfig org will contain both the namespace prefixes and their properties, the discussion about all of them will inevitably happen here and responsibilities will be hard to split.

**Comment by cxw42** [comment]
I won't stand in the way of that.  However, I do think we could provide a valuable service by hosting the properties registry.  That would give users a single place to find the information.  I would require complete proposals; a community could draft a proposal among itself in its own venue before submitting it to EditorConfig.

**Comment by florianb** [comment]
Since the "how" to potentially provide information about third party properties is not directly tied to the basic intent of the proposal we could anyways start the vote.

I would like to see an aggregation service under the hood of EC. And I can imagine allowing third-party developers to "merge" in a reference to their external spec file in a first step.

In a second step i could imagine a bot crawling the reference-registry creating MR's for each reference's upstream change.

Then we approve/decline simply on base of our code of conduct.

**Comment by ppalaga** [comment]
OK, I like the idea of having two separate things under the EC org: 

1. Prefix registry
2. Aggregated property catalog - an aggregated content from all third party property catalogs whose URLs would be registered in 1. and whose schema would be defined by the EC spec.

**Comment by jontsai** [comment]
Hi there, I'd like to help out with EC. What's the best way to get involved?

**Comment by cxw42** [comment]
@jontsai Welcome!  I suggest opening a new issue and letting us know what your interests are.  If you have a Windows machine available, I could use help with debugging the Windows build failure on https://github.com/editorconfig/editorconfig-vim/pull/148 :) .

**Comment by Mpdreamz** [comment]
I'm down with 

Prefix registry that lists

1. prefix + owner (@github handler).
2. Link to off-site documentation. 

I see little benefit in forcing the properties to be explicitly listed in a single place, domain specific properties being loose is one of the reasons many folks are starting to rely on .editorconfig since it's a file usually already present in codebases . 

I do think we need to reserve the right to the board to transfer ownership (on top of first come first serve) in case a prefix's owner goes stale. This needs to be a very public process (requester opening a GH process) with sufficient baking time to transfer IF the board agrees.

**Comment by bmartin** [comment]
To add another use case, I intend to leverage `.editorconfig` files for some internal purposes. The syntax is great and just begging for it. My plan is to have tooling that uses `MY_COMPANY_INTENAL_ACRHONYM/INTERNAL_PROPERTY`. For example:

```ini
# The source files in this folder need up-to-date copyright lines
[*.cpp]
MY_COMPANY_ACRONYM/copyright_required = true
```

I would never wish to register that string with any third party. I would hope that any `editorconfig` parser would just blindly return that property so I could check if it was set. This would allow anybody to leverage the format in a really clean way.

**Comment by florianb** [comment]
Hi @bmartin - this should already be possible. You can use undefined namespaces to invent your own additions. It should be save to prefix your rule like this:

`com_mycorp__copyright_required = true`

You should then be able to use on of the editorconfig parsers to extract your rule.

**Comment by JoshMcCullough** [comment]
> I currently don't see a strong requirement justifying the investment of resources in creating and maintaining third-party namespaces. This is something third-party developers can perfectly take care for.

@florianb just to clarify, I don't think the goal is for you all to standardize third-party e.g. `ij_*`, but rather language-based options e.g. `ts_*` for TypeScript, `cs_*` for C#.

**Comment by florianb** [comment]
Thank you @JoshMcCullough - i am really open to what is being specified using EditorConfig and even for languages i would recommend starting with a domain based prefix. [Rosetta Code](https://www.rosettacode.org) counts 922 languages at the time of writing, i think for a decision to foster a list of prefixes there should be relatable goals and a perspective on what amount of work to expect.

**Comment by cxw42** [comment]
This was bumped again as #516, requesting we recommend a format for the names of custom keys.  Just thinking about the form of the keys, looking back through this issue and that one, I see several proposals.  

I'd like to take an informal poll.  Would those reading this please react to this post to vote for one or more of the following? ---
|Reaction|Option|
|---|---|
| :+1: | `com_mycorp__property` |
| :smile: | `com.mycorp.property` |
| :tada: | `mycorp.com/property` |
| :heart: | None of the above |

I did some quick tests, and the C core can handle keys with dots and slashes just fine: 
[0001-Add-tests-of-keys-containing-dots-or-slashes-in-thei.txt](https://github.com/user-attachments/files/17122293/0001-Add-tests-of-keys-containing-dots-or-slashes-in-thei.txt) (produced with `git-format-patch`)

**Comment by florianb** [comment]
Let's push this forward..

**Comment by sdavids** [comment]
### Random thoughts

Most programming languages disallow periods and slashes in identifiers.

Mapping the keys to struct/record/object fields therefore needs additional language/framework specific work.

This can be seen as a pro or con.

---

[Snake case](https://en.wikipedia.org/wiki/Snake_case) has the advantage that one can double click the key to select the entire key; this works in the browser and almost all editors and terminals.

A key with dots can be double clicked in the browser but in most editors and terminals the period is considered a word separator.

Having a slash prohibits double clicking in all environments.

Double clicking to select the entire key is useful when one wants to copy the key from documentation into `.editorconfig`—less motor skills involved, no possibility of typos.

---

The spec does not [restrict the key length](https://spec.editorconfig.org/#id11).

Having a reverse-domain-name setup [on average](https://domainwheel.com/domain-name-statistics/) leads to prefix length of 11 characters (`com.sixsix.`).

That is 1/8 of the 80-column limit a lot of people prefer.

It gets worse when one considers [internationalized domain names](https://en.wikipedia.org/wiki/Punycode).

Also, if there is more than one product under the domain name then one will get even longer prefixes:

`com.jetbrains.fleet`
`com.jetbrains.resharper`

---
If we look at [some existing keys](https://www.jetbrains.com/help/resharper/EditorConfig_Index.html)…

Boy, are they long even without a prefix:

`blank_lines_after_file_scoped_namespace_directive`
`space_between_method_declaration_empty_parameter_list_parentheses`

Not being able to double click to copy and paste would be 😩

---

Using domain based keys has the problem of domains (companies) being subsumed into new enties, i.e. companies buying companies.

Using product name based keys has the same issue when a product gets renamed (most often when a company buys another and rebrands stuff).

So here one could think about what is more stable: the product name or the company/organisation providing the product.

Anecdote: I have seen codebases with more than 5 different `import com.mycorp` all due to company mergers and buyouts 🤣

---

Let's think about scale.

How many editors/plugins are likely to introduce new keys?

[EditorConfig](https://editorconfig.org/#pre-installed)
[EditorConfig](https://editorconfig.org/#download)
[List of text editors](https://en.wikipedia.org/wiki/List_of_text_editors)

Are we talking about 1..10, 1..100, 1..1000, 1...every second a new editor gets born

---

My suggestion would be to just require a prefix and that the prefix ends in `__`.

Maybe require that the entire key is in (screaming) snake case.

Whether the prefix signifies a company/organisation, product, language, or framework is not specified.

**Comment by sdavids** [comment]
On the topic of "How do I know if a prefix is already taken?" (registry):

I suggest going with a low-tech, low-fuss solution until it does not scale anymore:

* Open a PR with the desired prefix (maybe in https://github.com/editorconfig/editorconfig)

* GH Action
** check the prefix against some regular expression
** check for uniqueness
** update the list
** publish to
https://editorconfig.org
https://spec.editorconfig.org
or a new site

This can be done fully automatic or if spam/prefix-squatting is a concern then introduce a manual step and/or require more than just the prefix.

**Comment by osdm** [comment]
Oh please don't do "com.orgname.productname" thing. It would be just unreadable, and nobody would use them, preferring to stick with existing non-standard way. I could only reiterate my preferences:
1. An ideal option would be to simply use one-underscore delimiter, so we could support existing widely used prefixes like "csharp_", "ij_", "dotnet_", "resharper_"
2. Otherwise, we may use dots to delimit prefixes, but that would be simple prefixes, granted to those who own a product with such name, not complex URL names, so it would be like "csharp." (should be granted to DotNet Foundation only), "intellij." (to JetBrains), etc.

**Comment by florianb** [comment]
I really understand the points made - but i can tell some frustration about the EditorConfig comes from the fact that everything is being discussed for years without consent about how to move on.

In my opinion we're here to allow a pragmatic and backwards-compatible way to adopt cutting edge capabilities as they might be invented by any generous editor developer.

I'd prefer having the dot-notion versus snake-case just to reduce mental load when having the domain-based keys.

I prefer having the domain-notation since everybody is free to do in its domain whatever they want. I remember disliking the domain-notation in other parts of programming (Java?) but once your editoconfig is set it's done.

As soon a new feature seems to be mature in a sense of wider adoption i am happy to support making it a first class property.

And while i am not against a name registry per se - it is a origin of work, renaming, ownership, discussions. I would like to avoid this if not necessary.

**Comment by cxw42** [comment]
@sdavids your point about dot vs. slash to select words in a terminal is well-made --- that seems to me a point in favor of dot (or underscore, but `_` vs. `__` seems to me error-prone).

@florianb I agree with your point about the name registry.  @sdavids do you have time to do the work?  

Reverse-DNS is tedious to type, but easy to copy and paste.  It delegates the problem of prefix uniqueness to an existing system that works well (Internet DNS).

**Comment by osdm** [comment]
@cxw42 
> Reverse-DNS is tedious to type, but easy to copy and paste. It delegates the problem of prefix uniqueness to an existing system that works well (Internet DNS).

The problem is not that it is tedious to type, but that it is very hard to read. It would just be a nightmare.

**Comment by cxw42** [comment]
@osdm Reverse-DNS doesn't seem that bad to me, since the most relevant information is right next to the equal sign.   However, I respect that you and I disagree on this point.

In my experience, I write way more `.editorconfig` files than I read.  After all, I have `.editorconfig` files specifically so I don't have to care about what the settings are :D .  I don't personally think human-readability of a `.editorconfig` file is important enough to determine the key format.

**Comment by cxw42** [comment]
I just ran across another example of a dotted key in the wild --- https://github.com/dotnet/runtime/blob/83b0d939bedadf7d782b0b26307c2d8c1d5b76f4/.editorconfig#L55
```
dotnet_naming_rule.constant_fields_should_be_pascal_case.severity = suggestion
```

**Comment by florianb** [comment]
Should we collect the board's opinion? I don't see any further arguments that may pop up.

**Comment by xuhdev** [comment]
> Should we collect the board's opinion? I don't see any further arguments that may pop up.

Yes, I agree. But we still need to be clear on what to vote for.

Based on discussion, we don't necessarily need the reversed DNS syntax. We can just say put a unique identifier before a dot at the editor's discretion. Multiple dots are allowed so some editor's may choose to use reversed DNS style identifier but it's not required.

**Comment by ffes** [comment]
> But we still need to be clear on what to vote for.

Very important indeed.

We should be aware that there are already quite some prefixes out there and they are **very unlikely** to change. IMHO the definition in the specs should just be clear that a prefix is required.

Even if it is a dot or an underscore after the domain is a hard one for me, because those existing prefixes all use an underscore. Why should we "suddenly" decide it should be a dot?

And the specs should be clear what "domain specific properties" means and when to use them.

Less important at the moment, but what about [section ideas in the wiki](https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#ideas-for-domain-specific-properties)? Some of the suggested properties in the wiki are domain specific, others are less so. Should we even suggest domain specific properties?

**Comment by florianb** [comment]
Absolutely @ffes - just to make it clear, i think this won't be more than a recommendation. For me this issue isn't about making EC more strict and suppress any misaligned properties.

I would like to ensure that the introduction of custom properties can be done in a future safe way.

Recommending dot separation reduces the risk of future conflicts with the spec, having domain related properties may reduce potential conflicts between tool creators.

This spec change will have more of a directing character for custom properties.

In regard of section properties i'd like to discuss this separately because i think then we must talk about the current border of language-awareness as well.

**Comment by Mpdreamz** [comment]
> Even if it is a dot or an underscore after the domain is a hard one for me, because those existing prefixes all use an underscore. Why should we "suddenly" decide it should be a dot?

I subscribe to this line of thought too @ffes. Prior art trumps theory in this case for me. 

Prefixes are suffixed with `_` 

EditorConfig standard key/values:

```
{key_name) = ...
```

Domain key/values

```
{domain}_{key_name} = ...
```

In the unlikely event these clash `Editorconfig standard keys` always take priority and we specify domain implementations should take the breaking change.

If we can agree to this I am happy to submit an initial PR to the specification?


We can then follow up with how to deal with the registry as a separate issue to not get stuck.

**Comment by Mpdreamz** [comment]
> I would like to ensure that the introduction of custom properties can be done in a future safe way.
> Recommending dot separation reduces the risk of future conflicts with the spec, having domain related properties may reduce potential conflicts between tool creators.

I think this is unlikely given most tool vendors have already taken the approach to have their own `prefix_`. Part of the spec would advise tool vendors to never create unprefixed keys.

I agree (again:)) with @ffes that these section ideas: https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#ideas-for-domain-specific-properties are hurtful and most if not all are not generic enough to live in the unprefixed space.

**Comment by florianb** [comment]
I agree and i brought my points to attention, i am completely fine to be overruled. My goal is to bring this forward. :)

I am also open to discuss a registry in a separate issue.

**Comment by xuhdev** [comment]
Still doesn't seem clear what we would like to vote on yet :( we can also consider a double underscore separation.

I agree we need to move forward somehow...

**Comment by cxw42** [comment]
I don't think we need to worry about the existing custom properties since this is advisory, not mandatory.  I suggest we ask what we would recommend if we were having this conversation in a hypothetical world without existing custom properties.

Having thought about this for a while, I personally care more about the way we avoid conflicts than about, e.g., dot vs. underscore.  I think if all properties are unique then editors can unambiguously support whatever custom properties they wish.

**Comment by treyhunner** [comment]
> IMHO the definition in the specs should just be clear that a prefix is required.

This is along my lines of thought, preferably with some ideas for possible prefixes, notes of prior art, and our recommendation (which for me would be `something_likely_unique_`).

We would eventually want someone to write something up, but I don't think we're even at a consensus on what the points we'll want to note in the recommendation might be. Maybe we need a (whether formal or informal) vote to poll current feelings before the eventual write-up vote?
